tar/private/tar.bzl

```diff
@@ -474,29 +496,49 @@ def _mtree_impl(ctx):
             # copy workspace name here just in case to prevent ctx
             # to be transferred to execution phase.
             workspace_name = str(ctx.workspace_name)
+            format_each = "{}/%s".format(runfiles_dir)
 
-            content.add(_mtree_line(runfiles_dir, type = "dir"))
+            content.add(_mtree_line(runfiles_dir + "/", type = "dir"))
             content.add_all(
                 s.default_runfiles.empty_filenames,
-                format_each = "{}/%s".format(runfiles_dir),
+                format_each = format_each,
                 # be careful about what you pass to map_each as it will carry the data structures over to execution phase.
-                map_each = lambda f, e: _mtree_line(_vis_encode(f.removeprefix("external/") if f.startswith("external/") else workspace_name + "/" + f), "file"),
+                map_each = lambda f, e: _mtree_line(_vis_encode(f.removeprefix("../") if f.startswith("../") else workspace_name + "/" + f), "file"),
                 allow_closure = True,
             )
             content.add_all(
                 s.default_runfiles.files,
-                expand_directories = True,
                 uniquify = True,
-                format_each = "{}/%s".format(runfiles_dir),
+                format_each = format_each,
+                # be careful about what you pass to map_each as it will carry the data structures over to execution phase.
+                map_each = lambda f, e: _expand(f, e, _to_rlocation_path(f, workspace_name)),
+                allow_closure = True,
+            )
+            content.add_all(
+                s.default_runfiles.symlinks,
+                uniquify = True,
+                format_each = format_each,
                 # be careful about what you pass to map_each as it will carry the data structures over to execution phase.
-                map_each = lambda f, e: _expand(f, e, lambda f: _to_rlocation_path(f, workspace_name)),
+                map_each = lambda s, e: _expand(s.target_file, e, paths.normalize(workspace_name + "/" + s.path)),
                 allow_closure = True,
             )
+            content.add_all(
+                s.default_runfiles.root_symlinks,
+                uniquify = True,
+                format_each = format_each,
+                map_each = _expand_root_symlink,
+            )
+
             if repo_mapping != None:
                 content.add(
                     _mtree_line(_vis_encode(runfiles_dir + "/_repo_mapping"), "file", content = _vis_encode(repo_mapping.path)),
                 )
 
+            nested_directories = []
+            nested_directories.extend([s.target_file for s in s.default_runfiles.symlinks.to_list() if s.target_file.is_directory])
```

Comment:
i don't understand what's happening here; is it due to something like https://github.com/bazelbuild/bazel/issues/11811?